### ISSUE TITLE
coord: improve index availability with dataflow builder

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#8318]
#8241 demonstrated two different bugs when sequence_timestamp and the dataflow builder differ when deciding the set of available indexes. We've worked around the known bugs for now, but we should improve this situation by deciding on a way to communicate this information

**Comment by mjibson** [comment]
#8021 is also this problem.

My current thoughts: something needs to make a decision about usable indexes and tell that to something else. Candidates for things include: determine timestamp, dataflow builder, optimizer, transactions. A proposal on how to order things:

1. Determine a set of SQL objects (tables, views, sources) to care about. For one-off query this is just the things in the query (i.e., do nothing extra). For a transaction it includes all nearby SQL objects.
2. Determine a set of indexes those SQL objects depend on. If in a transaction, write that down.
3. Choose a timestamp based on those indexes. If in a transaction, write that down.
4. Give the expr to the optimizer along with this set of indexes that it could choose from.
5. Build a dataflow based on those indexes/optimized expr.

Today we approximately do 1-3, but not 4 or 5 (the optimizer and dataflow builder are free to use any index they want). I think 5 we can just do and there's not anything interesting about it.

4 concerns me because I don't know anything about the optimizer. I'm concerned about the case where step 2 figures out some indexes, but the optimizer would have liked to use another, better index, but it can't because it's not available. How can we guarantee that the set of indexes we pass to the optimizer are always the optimal set of indexes? I'm worried about a case where maybe there's some select over `V1`, and `V1` has an index, but `V1` depends on `V2` which is also indexed. Is there a case where the optimizer would ever want to use `V2` due to something about the query, even though the query specifies `V1`?

For the bootstrap bug in #8021, 5 would have changed the dataflow builder api to take an expression or set of indexes, and so that bootstrap problem would also get fixed here since bootstrap would have to communicate that information.

**Comment by frankmcsherry** [comment]
Re 4., the coordinator calls the optimizer with a list of indexes it is allowed to use. E.g.:
```rust
            let optimized_expr = self
                .view_optimizer
                .optimize(source.expr, self.catalog.enabled_indexes())?;
```

**Comment by mjibson** [comment]
Oh great TIL!

**Comment by mjibson** [comment]
I'm not convinced this should have been closed. The problem wasn't indexes on a compute instance, it was indexes in a transaction that are useable due to read holds.

**Comment by benesch** [comment]
I think it should be fine now? This code has been refactored a bunch in the past three days:

https://github.com/MaterializeInc/materialize/blob/6bd076e2d876e9f06ee50da8950d0d7308f5f3e0/src/coord/src/coord.rs#L2888-L2904

The new flow for each query is:

 1. If a transaction already has a timestamp, skip to step 5.
 2. Construct the ID bundle for a transaction's time domain based on the objects the query references using all known indexes.
 3. Determine a timestamp based on that ID bundle.
 4. Store that ID bundle in the transaction's `ReadHolds` struct.
 5. Construct the ID bundle for the query using all known indexes. If it is not a subset of the ID bundle in the `ReadHolds`, complain.
 6. Optimize the query using all known indexes. 

I believe step 5 is what provides the safety. If, say, a new index on a view was added _after_ the transaction started, and that index would be used by the optimizer in step 6 even though there's no read hold on that index, the query will be properly rejected by step 5.

**Comment by mjibson** [comment]
Hmm ok. I thought there was still a thing in opt/dataflow/render that assumed it could use all indexes instead of limiting itself to specified indexes, but maybe that isn't true or got fixed.

**Comment by mjibson** [comment]
Pertinent discussion: https://github.com/MaterializeInc/materialize/discussions/8386#discussioncomment-1829258

**Comment by benesch** [comment]
> Hmm ok. I thought there was still a thing in opt/dataflow/render that assumed it could use all indexes instead of limiting itself to specified indexes, but maybe that isn't true or got fixed.

Right, the optimizer assumes it can use _all_ indexes on a compute instance; not just those on which the transaction has acquired read holds. But the transaction code knows that now, and will refuse to plan a query if the optimizer _could_ reference an index on which the transaction did not acquire a read hold.
